iroh-rpc-client/src/status.rs

```diff
@@ -0,0 +1,92 @@
+    HealthCheckResponse,
+};
+
+#[tracing::instrument(skip(health_client))]
+pub async fn check(health_client: HealthClient<Channel>, service: String) -> ServiceStatus {
+    let req = iroh_metrics::req::trace_tonic_req(HealthCheckRequest { service });
+    let res = health_client.clone().check(req).await;
+    match res {
+        Ok(res) => res.into_inner().into(),
+        Err(s) => ServiceStatus::Down(s),
+    }
+}
+
+#[tracing::instrument(skip(health_client))]
+pub async fn watch(
+    health_client: HealthClient<Channel>,
+    service: String,
+) -> impl Stream<Item = ServiceStatus> {
+    stream! {
+        loop {
+            let req = iroh_metrics::req::trace_tonic_req(HealthCheckRequest { service: service.clone() });
+            let res = health_client.clone().watch(req).await;
+            match res {
+                Ok(stream) => {
+                    let mut stream = stream.into_inner();
+                    // loop over the stream, breaking if we get an error or stop receiving messages
+                    loop {
+                        match stream.message().await {
+                            Ok(Some(message)) => yield message.into(),
+                            Ok(None) => {
+                                yield ServiceStatus::Down(tonic::Status::new(tonic::Code::Unavailable, format!("No more health messages from service `{}`", service)));
+                                break;
+                            }
+                            Err(status) => {
+                                yield ServiceStatus::Down(status);
+                                break;
+                            }
+                        }
+                    }
+                },
+                Err(status) => yield ServiceStatus::Down(status)
+            }
+            /// wait before attempting to start a watch stream again
+            tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
+        };
+    }
+}
+
+impl std::convert::From<HealthCheckResponse> for ServiceStatus {
+    fn from(h: HealthCheckResponse) -> Self {
+        match h.status() {
+            ServingStatus::Unknown => ServiceStatus::Unknown,
+            ServingStatus::Serving => ServiceStatus::Serving,
+            ServingStatus::NotServing => ServiceStatus::NotServing,
+            ServingStatus::ServiceUnknown => ServiceStatus::ServiceUnknown,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum ServiceStatus {
+    ///  Indicates rpc server is in an unknown state
+    Unknown,
+    /// Indicates service is serving data
+    Serving,
+    /// Indicates service is not serving data, but the rpc server is not down
+    NotServing,
+    /// Indicates that the requested service is unknown
+    ServiceUnknown,
+    /// Indicates that the service is down. This ServiceStatus is assigned when
+    /// a `check` or `watch` call has returned an error with `tonic::Status`
+    Down(tonic::Status),
+}
+
+impl std::clone::Clone for ServiceStatus {
```

Comment:
why the need for a manual impl?

Reply:
`tonic::Status` does not implement `Clone` or `PartialEq`, and `ServiceStatus::Down()` contains a `tonic::Status`

Open to alternatives!

Reply:
might be good to file an issue

Reply:
with tonic